src/EspressoSGXTEEVerifier.sol

```diff
@@ -0,0 +1,179 @@
+ */
+contract EspressoSGXTEEVerifier is IEspressoSGXTEEVerifier, Ownable2Step {
+    using BytesUtils for bytes;
+
+    // V3QuoteVerififer contract from automata to verify the quote
+    V3QuoteVerifier public quoteVerifier;
+
+    mapping(bytes32 => bool) public registeredEnclaveHash;
+    mapping(bytes32 => bool) public registeredEnclaveSigner;
+    mapping(address => bool) public registeredSigners;
+
+    constructor(bytes32 enclaveHash, bytes32 enclaveSigner, address _quoteVerifier)
+        Ownable(msg.sender)
+    {
+        quoteVerifier = V3QuoteVerifier(_quoteVerifier);
+        registeredEnclaveHash[enclaveHash] = true;
+        registeredEnclaveSigner[enclaveSigner] = true;
+    }
+
+    /*
+        @notice Verify a quote from the TEE and attest on-chain
+        The verification is considered successful if the function does not revert.
+        @param rawQuote The quote from the TEE
+        @param reportDataHash The hash of the report data
+    */
+    function verify(bytes calldata rawQuote, bytes32 reportDataHash)
+        public
+        view
+        returns (EnclaveReport memory)
+    {
+        // Parse the header
+        Header memory header = parseQuoteHeader(rawQuote);
+
+        // Currently only version 3 is supported
+        if (header.version != 3) {
+            revert InvalidHeaderVersion();
+        }
+
+        // Verify the quote
+        (bool success,) = quoteVerifier.verifyQuote(header, rawQuote);
+        if (!success) {
+            revert InvalidQuote();
+        }
+
+        // Parse enclave quote
+        uint256 lastIndex = HEADER_LENGTH + ENCLAVE_REPORT_LENGTH;
+        EnclaveReport memory localReport;
+        (success, localReport) = parseEnclaveReport(rawQuote[HEADER_LENGTH:lastIndex]);
+        if (!success) {
+            revert FailedToParseEnclaveReport();
+        }
+
+        // Check that mrEnclave match
+        if (!registeredEnclaveHash[localReport.mrEnclave]) {
+            revert InvalidEnclaveHash();
+        }
+
+        if (!registeredEnclaveSigner[localReport.mrSigner]) {
+            revert InvalidEnclaveSigner();
+        }
+
+        //  Verify that the reportDataHash if the hash signed by the TEE
+        // We do not check the signature because `quoteVerifier.verifyQuote` already does that
+        if (reportDataHash != bytes32(localReport.reportData.substring(0, 32))) {
+            revert InvalidReportDataHash();
+        }
+
+        return localReport;
+    }
+
+    /*
+        @notice Register a new signer by verifying a quote from the TEE
+        @param attestation The attestation from the TEE
+        @param data which the TEE has attested to
+    */
+    function registerSigner(bytes calldata attestation, bytes calldata data) external {
+        // Check that the data length is 20 bytes because an address is 20 bytes
+        if (data.length != 20) {
+            revert InvalidDataLength();
+        }
+        bytes32 paddedSignerAddress = keccak256(data);
```

Comment:
checking my understanding and following up with a question.
so in this function we are adding a signer which is represented as a regular address, it's passed in as bytes because that's the output of the TEE. Why do we need to hash the data again? Is it because the enclave signed `keccak256(data)`?

Reply:
Yes, the enclave signed the hash of the the data ( data is the the bytes representation of the address)